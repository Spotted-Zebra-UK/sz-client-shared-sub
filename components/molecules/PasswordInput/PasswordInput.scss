@import '../../../variables';

.PasswordInput {
  border-bottom: 2px solid $black;
  display: flex;

  input {
    border: 0;
    font-size: $p1-font-size;
    height: 36px;
    line-height: $medium-line-height;
    width: calc(100% - 22px);
    font-size: 14px;
  }

  &__Button {
    border: 0;
    background-color: transparent;
    cursor: pointer;

    svg {
      width: 22px;

      path {
        fill: rgba(0, 0, 0, 0.16);
      }
    }
  }
}
