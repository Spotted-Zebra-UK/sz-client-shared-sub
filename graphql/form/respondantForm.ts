import { gql } from '@apollo/client';

// TODO: Refactor, use generated query instead of this.
export const RESPONDANT_FORM_QUERY = gql`
  query RespondantFormCompletedCheck(
    $associatedId: Int!
    $formOwnerId: Int!
    $formType: FormType!
    $respondantId: Int
  ) {
    respondantForm: RespondantForm(
      associatedId: $associatedId
      formOwnerId: $formOwnerId
      formType: $formType
      respondantId: $respondantId
    ) {
      isComplete
      submittedFormLogId
      formType
      formOwnerId
      associatedId
      formId
      respondantFieldWithAnswers {
        field {
          id
          question
          hint
          settings {
            allowFreeText
            searchable
          }
          isArchived
          fieldType
          ... on CompanyEmployeeSelectField {
            dynamicSelectOptions
          }
          ... on MultiSelectField {
            selectOptions
          }
          ... on SingleSelectField {
            selectOptions
          }
        }
        displayOrder
        isOptional
        answer
        selectOptionAnswerIsFreeText
      }
    }
  }
`;

// TODO: Refactor, use generated query instead of this.
export const RESPONDANT_FORM_UPDATE_MUTATION = gql`
  mutation RespondantFormUpdate(
    $associatedId: Int!
    $formOwnerId: Int!
    $formType: FormType!
    $answers: [UpdateRespondantFormAnswerArgs!]
  ) {
    respondantForm: RespondantFormUpdate(
      associatedId: $associatedId
      formOwnerId: $formOwnerId
      formType: $formType
      answers: $answers
    ) {
      isComplete
      submittedFormLogId
      formType
      formOwnerId
      associatedId
      formId
      respondantFieldWithAnswers {
        field {
          id
          question
          hint
          settings {
            allowFreeText
            searchable
          }
          isArchived
          fieldType
          ... on CompanyEmployeeSelectField {
            dynamicSelectOptions
          }
          ... on MultiSelectField {
            selectOptions
          }
          ... on SingleSelectField {
            selectOptions
          }
        }
        displayOrder
        isOptional
        answer
        selectOptionAnswerIsFreeText
      }
    }
  }
`;
