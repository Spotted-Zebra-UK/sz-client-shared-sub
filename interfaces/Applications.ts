export enum Application {
  CANDIDATE = 'candidate',
  COMPANY = 'company',
}
