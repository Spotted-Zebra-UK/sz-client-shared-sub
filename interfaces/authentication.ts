export interface IRequestTokenRefreshResponse {
  requestTokenRefresh: {
    accessToken: string;
    refreshToken: string;
  };
}

export interface IRequestTokenRefreshInput {
  accessToken: string;
  refreshToken: string;
}

export interface IDeauthenticateInput {
  accessToken: string;
}
