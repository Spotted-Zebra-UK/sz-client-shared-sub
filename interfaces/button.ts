export type TButtonType = 'submit' | 'button' | 'link';
