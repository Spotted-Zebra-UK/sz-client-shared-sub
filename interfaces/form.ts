export type TTextFromFieldValue = string;
export type TSelectFormFieldOptionValue = string | number | undefined;
export type TSelectFormFieldValue =
  | string
  | number
  | undefined
  | string[]
  | number[];
export type TCheckboxFormFieldValue = boolean;
export type TFormFieldValue =
  | TSelectFormFieldValue
  | TTextFromFieldValue
  | TCheckboxFormFieldValue;
