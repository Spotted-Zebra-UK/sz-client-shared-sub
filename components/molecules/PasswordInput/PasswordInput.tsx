import './PasswordInput.scss';
import React, { FC, useState } from 'react';
import { ReactComponent as EyeIcon } from '../../../icons/Eye.svg';
import { ReactComponent as EyeSolidIcon } from '../../../icons/EyeSolid.svg';
import Button from '../../atoms/Button/Button';

interface IPasswordInput {
  name: string;
  onChange: (value: string, name: string) => void;
  placeholder?: string;
  value: string;
}

const PasswordInput: FC<IPasswordInput> = ({
  value,
  onChange,
  placeholder,
  name,
}) => {
  const [isVisible, setIsVisible] = useState(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value, name);
  };

  const handleVisibilityChange = () => {
    setIsVisible(prevIsVisible => !prevIsVisible);
  };

  return (
    <div className="PasswordInput">
      <input
        onChange={handleChange}
        name={name}
        placeholder={placeholder}
        value={value}
        type={isVisible ? 'text' : 'password'}
      />
      <Button
        onClick={handleVisibilityChange}
        className="PasswordInput__Button"
      >
        {isVisible ? <EyeSolidIcon /> : <EyeIcon />}
      </Button>
    </div>
  );
};

export default PasswordInput;
