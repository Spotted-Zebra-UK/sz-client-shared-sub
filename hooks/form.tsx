import { useState } from 'react';
import validate from '../helpers/validate';
import { TFormFieldValue } from '../interfaces/form';

export type TFormValues = {
  [key in string]: TFormFieldValue;
};

export type TFormErrors = {
  [key in string]: string;
};

interface IUseFormOptions {
  initialValues: TFormValues;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  validationShema?: any;
}

interface IUseFormProps {
  onSubmit: (values: TFormValues) => void;
  options: IUseFormOptions;
}

export const useForm = ({
  options,
  onSubmit,
}: IUseFormProps): [
  TFormValues,
  TFormErrors | undefined,
  (newValue: TFormFieldValue, name: string) => void,
  (
    event:
      | React.FormEvent<HTMLFormElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
      | undefined
  ) => void
] => {
  const [values, setValues] = useState<TFormValues>(options.initialValues);
  const [errors, setErrors] = useState<TFormErrors | undefined>();

  const validateForm = () => {
    if (!options.validationShema) {
      return undefined;
    }

    return validate(values, options.validationShema);
  };

  const handleChange = (newValue: TFormFieldValue, name: string) => {
    setValues(prevValues => ({ ...prevValues, [name]: newValue }));

    if (errors && errors[name]) {
      setErrors(prevErrors => {
        return { ...prevErrors, [name]: '' };
      });
    }
  };

  const handleSubmit = (
    event:
      | React.FormEvent<HTMLFormElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
      | undefined
  ) => {
    event?.preventDefault();
    const newErrors = validateForm();
    if (!newErrors) {
      onSubmit(values);
    } else {
      setErrors(newErrors);
    }
  };

  return [values, errors, handleChange, handleSubmit];
};
