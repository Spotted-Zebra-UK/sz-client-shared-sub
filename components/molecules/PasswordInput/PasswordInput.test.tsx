import { mount, shallow } from 'enzyme';
import { ReactComponent as EyeIcon } from '../../../icons/Eye.svg';
import { ReactComponent as EyeSolidIcon } from '../../../icons/EyeSolid.svg';
import { TTextFromFieldValue } from '../../../interfaces/form';
import PasswordInput from './PasswordInput';

describe('PasswordInput', () => {
  let onChange: (value: TTextFromFieldValue, name: string) => void;

  const value = 'test-value';
  const name = 'test-name';

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render correctly with required props', () => {
    const wrapper = shallow(
      <PasswordInput name={name} value={value} onChange={onChange} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(EyeIcon)).toHaveLength(1);
    expect(wrapper.find(EyeSolidIcon)).toHaveLength(0);
    expect(wrapper.find('input').props().type).toEqual('password');
  });

  it('should change input type on visibility change button click', () => {
    const wrapper = mount(
      <PasswordInput name={name} value={value} onChange={onChange} />
    );

    wrapper.find('button').simulate('click');

    expect(wrapper.find(EyeIcon)).toHaveLength(0);
    expect(wrapper.find(EyeSolidIcon)).toHaveLength(1);
    expect(wrapper.find('input').props().type).toEqual('text');

    wrapper.find('button').simulate('click');

    expect(wrapper.find(EyeIcon)).toHaveLength(1);
    expect(wrapper.find(EyeSolidIcon)).toHaveLength(0);
    expect(wrapper.find('input').props().type).toEqual('password');
  });

  it('should call onChange on input change', () => {
    const newValue = 'test-newValue';
    const wrapper = shallow(
      <PasswordInput name={name} value={value} onChange={onChange} />
    );

    wrapper.find('input').simulate('change', { target: { value: newValue } });

    expect(onChange).toHaveBeenCalledWith(newValue, name);
  });

  it('should render correctly when placeholder provided', () => {
    const placeholder = 'test-placeholder';
    const wrapper = shallow(
      <PasswordInput
        placeholder={placeholder}
        name={name}
        value={value}
        onChange={onChange}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
