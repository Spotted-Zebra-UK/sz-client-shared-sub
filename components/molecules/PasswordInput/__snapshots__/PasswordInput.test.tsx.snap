// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should render correctly when placeholder provided 1`] = `
<div
  className="PasswordInput"
>
  <input
    onChange={[Function]}
    placeholder="test-placeholder"
    type="password"
    value="test-value"
  />
  <Button
    className="PasswordInput__Button"
    onClick={[Function]}
  >
    <ForwardRef(SvgEye) />
  </Button>
</div>
`;

exports[`PasswordInput should render correctly with required props 1`] = `
<div
  className="PasswordInput"
>
  <input
    onChange={[Function]}
    type="password"
    value="test-value"
  />
  <Button
    className="PasswordInput__Button"
    onClick={[Function]}
  >
    <ForwardRef(SvgEye) />
  </Button>
</div>
`;
